pkg/versioned/map.go

```diff
@@ -0,0 +1,170 @@
+// Copyright 2018 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package versioned
+
+import (
+	"github.com/cilium/cilium/pkg/lock"
+)
+
+// UUID is the UUID for the object that is going to be stored in the map.
+type UUID string
+
+// Map maps a UUID to an Object.
+type Map map[UUID]Object
+
+// NewMap returns an initialized Map.
+func NewMap() Map {
+	return make(Map)
+}
+
+// Add maps the uuid to the given obj.
+func (m Map) Add(uuid UUID, obj Object) {
+	m[uuid] = obj
+}
+
+// Delete deletes the value that maps uuid in the map. Returns true of false
+// if the object existed in the map before deletion.
+func (m Map) Delete(uuid UUID) bool {
+	_, exists := m[uuid]
+	if exists {
+		delete(m, uuid)
+	}
+	return exists
+}
+
+// Get returns the object that maps to the given uuid and returns true or false
+// either the object exists or not.
+func (m Map) Get(uuid UUID) (Object, bool) {
+	o, exists := m[uuid]
+	return o, exists
+}
+
+// DeepEqualFunc should return true or false if both interfaces `o1` and `o2`
+// are considered equal.
+type DeepEqualFunc func(o1, o2 interface{}) bool
+
+// ComparableMap is a map that can store Objects that are comparable between
+// each other.
+type ComparableMap struct {
+	Map
+	DeepEquals DeepEqualFunc
+}
+
+// NewComparableMap returns an initialized map with the equalFunc set as the
+// DeepEquals of the map.
+func NewComparableMap(equalFunc DeepEqualFunc) *ComparableMap {
+	return &ComparableMap{
+		Map:        NewMap(),
+		DeepEquals: equalFunc,
+	}
+}
+
+// AddEqual maps `uuid` to `newObj` if the object to be inserted has a newer
+// Version than the one already mapped in the map. Returns false if the object
+// inserted does is not mapped yet or if the object has a newer version and
+// is not deep equaled than the object already stored.
+func (m *ComparableMap) AddEqual(uuid UUID, newObj Object) bool {
```

Comment:
This function appears to be used directly in `replaceFuncFactory()` for the checking functionality, but it has side-effects on the existing map. Is that intended?

Reply:
Talked off-line I'll take a look at this.